### ISSUE TITLE
Validation does not occur on serialization documentation issue

### ISSUE THREAD

**Issue by matejsp** [marshmallow-code/marshmallow#2189]
Base on the documentation that validation does not occur on serialisation anymore  https://marshmallow.readthedocs.io/en/stable/upgrading.html.

There is example how to workaround. However this alternative is flawed since it does not work if you use data_key on created_at (like camlcase). 

Another case that does not work is if you actually pass a correct value (like current datetime).

Is there any alternative that takes data_key into account?
I think documentation would need to be updated to reflect that use case.

```
from marshmallow import Schema, fields, ValidationError

invalid_data = dict(created_at="invalid")


class WidgetSchema(Schema):
    created_at = fields.DateTime(data_key='createdAt')


# 2.x
# WidgetSchema(strict=True).dump(invalid_data)
# marshmallow.exceptions.ValidationError: {'created_at': ['"invalid" cannot be formatted as a datetime.']}

# 3.x
# WidgetSchema().dump(invalid_data)
# AttributeError: 'str' object has no attribute 'isoformat'

# Instead, validate before dumping
schema = WidgetSchema()
try:
    widget = schema.load(invalid_data)
except ValidationError as e:
    print("handle errors... {}".format(e.messages))
else:
    dumped = schema.dump(widget)
```

Another case that does not work:
```
invalid_data = dict(created_at=datetime.datetime.now())
```

**Comment by deckar01** [comment]
If you can't guarantee that data being `dump()`ed is valid, it needs to be `load()`ed.

https://marshmallow.readthedocs.io/en/stable/quickstart.html#deserializing-objects-loading

**Comment by matejsp** [comment]
Yes this is possible but is not the "alternative" according to docs. Better way would be do to validate() instead of load().

It says validate before dumping to get the same behaviour.
Perhaps docs should be changed to reflect this.

**Comment by nightcityblade** [comment]
Hi, I'd like to work on this. I'll submit a PR shortly.

**Comment by lafrech** [comment]
Loading before dumping only works on simple cases (dict, not obj, no data_key, etc.), so saying validate/load before dumping is wrong in the general case.

My take on this is that we should require objects to be valid and not try to provide a workaround.

The example added in #3030 doesn't illustrate a real-life scenario IMHO.

I'd just remove the example starting at "Instead" and rephrase as 

> [Schema.dump](https://marshmallow.readthedocs.io/en/stable/top_level.html#marshmallow.Schema.dump) will no longer validate and collect error messages. Data must be valid before being serialized.
